pkg/stats/combiner.go

```diff
@@ -28,55 +32,155 @@ type StatsObject interface {
 			}
-			return l.flush(now), nil
+		} else {
+			if err := item.Initialize(); err != nil {
+				return fmt.Errorf("failed to initialize object: %w", err)
+			}
+			existing = item
 		}
+
+		serialized, err := sc.serializer(existing)
+		if err != nil {
+			return fmt.Errorf("failed to serialize object: %w", err)
+		}
+		return bucket.Put(keyBytes, serialized)
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to update record: %w", err)
 	}
 
-	if err := item.Initialize(); err != nil {
-		return nil, err
+	// Flush if the time interval has passed
+	if now.After(sc.cutoff) {
+		flushedData, err = sc.flush(now)
+		if err != nil {
+			return nil, fmt.Errorf("failed to flush data: %w", err)
+		}
 	}
-	(*l.bucket)[key] = append((*l.bucket)[key], item)
-	return l.flush(now), nil
+
+	return flushedData, nil
 }
 
-func (l *StatsCombiner[T]) flush(now time.Time) *map[uint64][]T {
-	if now.Before(l.cutoff) {
-		return nil
+func (sc *StatsCombiner[T]) flush(now time.Time) (map[uint64][]T, error) {
+	flushedData := make(map[uint64][]T)
+	currentBucket := sc.getBucketName(sc.cutoff.Unix() - int64(sc.interval.Seconds()))
+	println("Flushing from bucket", currentBucket)
+
+	err := sc.db.Update(func(tx *bbolt.Tx) error {
+		bucket := tx.Bucket([]byte(currentBucket))
+		if bucket == nil {
+			return nil
+		}
+
+		err := bucket.ForEach(func(k, v []byte) error {
+			key := bytesToUint64(k)
+			item, err := sc.deserializer(v)
+			if err != nil {
+				return fmt.Errorf("failed to deserialize object: %w", err)
+			}
+			fmt.Printf("Flushing item: %v\n", string(v))
+			flushedData[key] = append(flushedData[key], item)
+			return nil
+		})
+		if err != nil {
+			return fmt.Errorf("failed to collect data: %w", err)
+		}
+
+		return tx.DeleteBucket([]byte(currentBucket))
+	})
+	if err != nil {
+		return nil, fmt.Errorf("flush failed: %w", err)
 	}
 
-	bucketpile := l.bucket
-	l.bucket = &map[uint64][]T{}
-	l.cutoff = now.Add(l.interval)
+	sc.cutoff = now.Truncate(sc.interval).Add(sc.interval)
 
-	return bucketpile
+	return flushedData, nil
 }
 
-func NewStatsCombiner[T StatsObject](now time.Time, interval time.Duration) *StatsCombiner[T] {
+func NewStatsCombiner[T StatsObject](
+	db *bbolt.DB,
+	bucketBase string,
+	now time.Time,
+	interval time.Duration,
+	serializer func(T) ([]byte, error),
+	deserializer func([]byte) (T, error),
+) *StatsCombiner[T] {
 	return &StatsCombiner[T]{
-		interval: interval,
-		cutoff:   now.Add(interval),
-		bucket:   &map[uint64][]T{},
+		interval:     interval,
+		cutoff:       now.Truncate(interval).Add(interval),
+		db:           db,
+		bucketBase:   bucketBase,
+		serializer:   serializer,
+		deserializer: deserializer,
 	}
 }
+
+func (sc *StatsCombiner[T]) currentBucketName(t time.Time) string {
+	return fmt.Sprintf("%s_%d", sc.bucketBase, t.Truncate(sc.interval).Unix())
+}
+
+func (sc *StatsCombiner[T]) getBucketName(t int64) string {
+	return fmt.Sprintf("%s_%d", sc.bucketBase, t)
+}
+
+func uint64ToBytes(v uint64) []byte {
```

Comment:
There is a binary package that will do some of this for you:

https://pkg.go.dev/encoding/binary#ByteOrder

There isn't a need to use it, but it will be easier than hand-rolling.